File: py/aoc2021/day12.py
from collections import defaultdict

SAMPLE_INPUT_1 = [
    "start-A",
    "start-b",
    "A-c",
    "A-b",
    "b-d",
    "A-end",
    "b-end",
]
SAMPLE_INPUT_2 = [
    "dc-end",
    "HN-start",
    "start-kj",
    "dc-start",
    "dc-HN",
    "LN-dc",
    "HN-end",
    "kj-sa",
    "kj-HN",
    "kj-dc",
]
SAMPLE_INPUT_3 = [
    "fs-end",
    "he-DX",
    "fs-he",
    "start-DX",
    "pj-DX",
    "end-zg",
    "zg-sl",
    "zg-pj",
    "pj-he",
    "RW-he",
    "fs-DX",
    "pj-RW",
    "zg-RW",
    "start-pj",
    "he-WI",
    "zg-he",
    "pj-fs",
    "start-RW",
]


def solve(lines, bonus):
    edges = defaultdict(set)
    for line in lines:
        lhs, rhs = line.strip().split("-", maxsplit=2)
        edges[lhs].add(rhs)
        edges[rhs].add(lhs)

    def walk(current, path, bonus):
        if current == "end":
            yield path
        else:
            for next in edges[current]:
                if next == "start":
                    pass
                elif next.isupper() or next not in path:
                    yield from walk(next, path + [next], bonus)
                elif bonus and next.islower():
                    yield from walk(next, path + [next], False)

    return sum(1 for _ in walk("start", ["start"], bonus))


def part1(lines):
    """
    >>> part1(SAMPLE_INPUT_1)
    10
    >>> part1(SAMPLE_INPUT_2)
    19
    >>> part1(SAMPLE_INPUT_3)
    226
    """
    return solve(lines, False)


def part2(lines):
    """
    >>> part2(SAMPLE_INPUT_1)
    36
    >>> part2(SAMPLE_INPUT_2)
    103
    >>> part2(SAMPLE_INPUT_3)
    3509
    """
    return solve(lines, True)


parts = (part1, part2)

File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2021/Day12.kt
package com.github.ephemient.aoc2021

class Day12(lines: List<String>) {
    private val start: Int
    private val end: Int
    private val edges: Map<Int, List<Pair<Int, Boolean>>>

    init {
        val names = mutableMapOf<String, Int>()
        val edges = mutableMapOf<Int, MutableList<Pair<Int, Boolean>>>()
        for (line in lines) {
            val (lhs, rhs) = line.split('-', limit = 2)
            val src = names.getOrPut(lhs) { names.size }
            val dst = names.getOrPut(rhs) { names.size }
            edges.getOrPut(src) { mutableListOf() }.add(dst to rhs.all { it.isUpperCase() })
            edges.getOrPut(dst) { mutableListOf() }.add(src to lhs.all { it.isUpperCase() })
        }
        check(names.size < Int.SIZE_BITS)
        start = names.getValue("start")
        end = names.getValue("end")
        this.edges = edges
    }

    private fun solve(bonus: Boolean): Int {
        var count = 0
        walk((if (bonus) 0 else Int.MIN_VALUE) to start) { (state, element) ->
            edges[element]?.mapNotNull { (next, big) ->
                when {
                    next == start -> null
                    next == end -> null.also { count++ }
                    big -> state to next
                    state.and(1 shl next) == 0 -> state.or(1 shl next) to next
                    state >= 0 -> state or Int.MIN_VALUE to next
                    else -> null
                }
            }.orEmpty()
        }
        return count
    }

    fun part1(): Int = solve(false)

    fun part2(): Int = solve(true)

    companion object {
        private fun <T> walk(start: T, step: (T) -> Iterable<T>) {
            val stack = mutableListOf(step(start).iterator())
            while (stack.isNotEmpty()) {
                val iterator = stack.last()
                if (!iterator.hasNext()) {
                    stack.removeLast()
                    continue
                }
                stack.add(step(iterator.next()).iterator())
            }
        }
    }
}
